cpmpy/solvers/gcs.py

```diff
@@ -0,0 +1,413 @@
+        super().__init__(name="Glasgow Constraint Solver", cpm_model=cpm_model)
+
+    def solve(self, time_limit=None, **kwargs):
+        """
+            Call the Glasgow Constraint Solver
+            Arguments:
+            - time_limit:  #TODO maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+            Additional keyword arguments:
+            - proof: filename/path for the two files ".opb" and ".veripb" necessary for proof verification
+            (if this argument is not supplied the proof files created by the solver are deleted).
+
+            Arguments that correspond to solver parameters:
+            #TODO document solver parameters (there are none at the moment)
+        """
+        # ensure all user vars are known to solver
+        self.solver_vars(list(self.user_vars))
+
+        if time_limit is not None:
+            raise NotImplementedError("Glasgow Constraint Solver does not currently support timeouts.")
+        
+        # call the solver, with parameters
+        gcs_stats = self.gcs.solve(**kwargs)
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        # self.cpm_status.runtime = self.gcs.time()
+
+        # translate exit status
+        if gcs_stats['solutions'] != 0:
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        else:
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return(self.cpm_status)
+
+        # translate solution values (of user specified variables only)
+        self.objective_value_ = None
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                sol_var = self.solver_var(cpm_var)
+                if isinstance(cpm_var, _BoolVarImpl):
+                    # Convert back to bool
+                    cpm_var._value = bool(self.gcs.get_solution_value(sol_var))
+                else:
+                    cpm_var._value = self.gcs.get_solution_value(sol_var)
+
+            # translate objective, for optimisation problems only
+            if self.has_objective:
+                self.objective_value_ = self.gcs.get_solution_value(self.objective_var)
+        
+        return has_sol
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        if is_num(cpm_var): # shortcut, eases posting constraints
+            return self.gcs.create_integer_constant(cpm_var)
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            # gcs only works with integer variables, so not(x) = -x + 1
+            return self.gcs.add_constant(self.gcs.negate(self.solver_var(cpm_var._bv)), 1)
```

Comment:
Also not sure about this?
`gcs.negate(x)` maps a new var id for the view `-x` (without creating a new underlying variable).
`gcs.add_constant(x, c)` similarly maps a new id for the view `x + c`.

So `~a|b` would be translated to `Or{view_of_a, b}` where `view_of_a` is a reference to `-a + 1` (we have a specific wrapper for viewed variables).
Since our `Or` constraint just asserts that at least one of it's arguments is non-zero this shouldn't cause any problems?